autoscale-workload-rabbit/src/main/java/com/github/autoscaler/workload/rabbit/RabbitWorkloadAnalyser.java

```diff
@@ -114,18 +135,61 @@ public ScalingAction analyseWorkload(final InstanceInfo instanceInfo)
         return ScalingAction.NO_ACTION;
     }
 
-
-    private int getWorkersNeeded(final long messages, final int backlogGoal, final InstanceInfo instanceInfo)
+    private int getWorkersNeeded(
+            final int messagesInTargetQueue,
+            final int messagesInStagingQueues,
+            final long messagesInTargetQueueAndStagingQueues,
+            final int backlogGoal,
+            final InstanceInfo instanceInfo,
+            final List<String> stagingQueueNames)
     {
-        double consume = statsQueue.stream().mapToDouble(QueueStats::getConsumeRate).average().getAsDouble();
-        double publish = statsQueue.stream().mapToDouble(QueueStats::getPublishRate).average().getAsDouble();
-        double avgMsgs = statsQueue.stream().mapToDouble(QueueStats::getMessages).average().getAsDouble();
+        double consume = targetQueueStatsQueue.stream().mapToDouble(QueueStats::getConsumeRate).average().getAsDouble();
+        double publish = targetQueueStatsQueue.stream().mapToDouble(QueueStats::getPublishRate).average().getAsDouble();
```

Comment:
The only thing publishing to the target queue is the message distributor (via shovels) so it is no longer relevant to a scaling decision. I think the publish rate of the staging queues is what we're interested in, do you agree?

Reply:
Think of it this way, before QoQ the publish rate indicated how quickly the backlog was growing.

Now the publish rate to what we call the target queue is curtailed by the target queue length limit meaning it's no longer an indicator of how quickly the backlog is growing and shouldn't contribute to the calculation you mention above.

If there was 1 staging queue with a publish rate of 1000 m/s and the target queue was full (implying a publish rate of 0 m/s) the spot average of new messages being published for this worker would be (1000 + 0)/2 = 500m/s? 

So I think the answer is "I'm not sure if we should change this to **only the publishing rate of the staging queues:**"

Reply:
Actually, do we have a bigger problem when calculating average publish rate? 
If there are many inactive staging queues (because that tenant hasn't been active for a while) won't the average will be diluted?

targetQueueStatsQueue.stream().mapToDouble(QueueStats::getPublishRate).average().getAsDouble();

Reply:
> Actually, do we have a bigger problem when calculating average publish rate? If there are many inactive staging queues (because that tenant hasn't been active for a while) won't the average be diluted?
> 
> targetQueueStatsQueue.stream().mapToDouble(QueueStats::getPublishRate).average().getAsDouble();

The average publishing rate is only calculated so that we can check that it is 0 (which will use along with other stats being 0 to indicate that we should scale the worker to 0):

```
else if ( Double.compare(0.0, targetQueueAvgConsumeRate) == 0 &&
                Double.compare(0.0, stagingQueuesAvgPublishRate) == 0 &&
                Double.compare(0.0, targetQueueAndStagingQueuesAvgMsgs) == 0 ) {
            // if we have no target queue consumption rate, no staging queues publish rate, and no messages then we don't need any workers
            return 0;
```

So inactive staging queues will dilute the rate, but they won't change it from > 0 to 0, and since we only ever compare the publishing rate to 0, I think thats ok?

Reply:
Thanks for the clarification Rory, sounds good to me.